### PR TITLE
reference/performance: add opt_rule_blacklist document

### PR DESCRIPTION
**Pull request by lzmhhh123**
### What is changed, added or deleted? (Required)
Add the document for how to use optimizer rule blacklist table.

close https://github.com/pingcap/tidb/issues/15207

### Which TiDB version(s) do your changes apply to? (Required)

- [X] master (the latest development version)
- [X] v4.0 (TiDB 4.0 versions)
- [X] v3.1 (TiDB 3.1 versions)
- [X] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from: https://github.com/pingcap/tidb/pull/11096

### CODE REVIEW

**Comment by TomShawn** [comment]
@lzmhhh123 Please involve a technical review.

File: reference/performance/opt-rule-blacklist.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+在[join reorder 算法简介](/reference/performance/join-reorder.md)中，介绍了一种较为复杂的优化规则，TiDB 对于某条查询，会智能地选择一些优化规则对查询进行智能优化。
+
+同时 TiDB 提供了一些可以自定义优化的功能，方便用户针对查询进行调优，比如[优化器 Hints](/reference/performance/optimizer-hints.md)。
+
+**优化规则黑名单**是针对优化规则的调优手段之一，主要用于手动禁用一些优化规则。
+
+## 重要的优化规则<div id="rules"></div>
+
+- 列裁剪（column_prune）：对于上层算子不要的列，裁剪掉下层算子中的该列，减少计算。
```
Comment (eurekaka) [inline reference/performance/opt-rule-blacklist.md:16]:
```suggestion
- 列裁剪（column_prune）：对于上层算子不需要的列，在下层算子不输出该列，减少计算。
```

File: reference/performance/opt-rule-blacklist.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+- 子查询去关联（decorrelate）：会尝试对自查询进行改写，将其转换为 join 或 aggression 计算。
```
Comment (eurekaka) [inline reference/performance/opt-rule-blacklist.md:18]:
```suggestion
- 子查询去关联（decorrelate）：会尝试对相关子查询进行改写，将其转换为普通 join 或 aggression 计算。
```

File: reference/performance/opt-rule-blacklist.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+- 聚合消除（aggregation_eliminate）：尝试删除执行计划中的某些聚合算子。
```
Comment (eurekaka) [inline reference/performance/opt-rule-blacklist.md:20]:
```suggestion
- 聚合消除（aggregation_eliminate）：尝试消除执行计划中的某些不必要的聚合算子。
```

File: reference/performance/opt-rule-blacklist.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+- 投影消除（projection_eliminate）：对于执行计划中不必要的投影算子，对其消除。
+
+- 最大最小消除（max_min_eliminate）：改写聚合中的 max, min 计算，转化为 order by + limit 1。
+
+- 谓词下推（predicate_push_down）：尝试将执行计划中过滤条件下推到更底层的节点。
```
Comment (eurekaka) [inline reference/performance/opt-rule-blacklist.md:26]:
```suggestion
- 谓词下推（predicate_push_down）：尝试将执行计划中过滤条件下推到离数据源更近的算子上。
```

File: reference/performance/opt-rule-blacklist.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+- 外连接消除（outer_join_eliminate）：尝试消除执行计划中不必要的 left join 或者 right join。
+
+- 分区表查询改写（partition_processor）：将分区表查询改成为用 union all 算子代替。
```
Comment (eurekaka) [inline reference/performance/opt-rule-blacklist.md:30]:
```suggestion
- 分区裁剪（partition_processor）：将分区表查询改成为用 union all，并裁剪掉不满足过滤条件的分区。
```

File: reference/performance/opt-rule-blacklist.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+- 聚合下推（aggregation_push_down）：尝试将执行计划中的聚合算子下推到更底层的计算节点。
```
Comment (eurekaka) [inline reference/performance/opt-rule-blacklist.md:32]:
```suggestion
- 聚合下推（aggregation_push_down）：尝试将执行计划中的聚合算子下推过 join 算子。
```

File: reference/performance/opt-rule-blacklist.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+- topN 下推（topn_push_down）：尝试将执行计划中的 topN 算子下推到更底层的计算节点。
```
Comment (eurekaka) [inline reference/performance/opt-rule-blacklist.md:34]:
```suggestion
- TopN 下推（topn_push_down）：尝试将执行计划中的 TopN 算子下推到离数据源更近的算子上。
```

File: reference/performance/opt-rule-blacklist.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+- join 重排序（join_reorder）：尝试对多个 join 算子进行重新排序。
```
Comment (eurekaka) [inline reference/performance/opt-rule-blacklist.md:36]:
```suggestion
- join 重排序（join_reorder）：对多表 join 确定连接顺序。
```

File: reference/performance/opt-rule-blacklist.md [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+## 如何禁用优化规则
+
+当某些优化规则对于一些比较特殊的查询，出现了不理想的优化结果时，用户可以使用**优化规则黑名单**对一些优化规则进行禁用。
+
+### 使用方法
+
+每个优化规则都有自己的名字，比如列裁剪的名字是 "column_prune"，所有优化规则的名字皆可以在[重要的优化规则](/reference/performance/opt-rule-blacklist.md#rules)的括号中查到。
```
Comment (eurekaka) [inline reference/performance/opt-rule-blacklist.md:44]:
```suggestion
每个优化规则都有自己的名字，比如列裁剪的名字是 "column_prune"，所有优化规则的名字皆可以在[重要的优化规则](/reference/performance/opt-rule-blacklist.md#rules)中查到。
```

File: reference/performance/opt-rule-blacklist.md [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+{{< copyable "sql" >}}
+
+```sql
+insert into mysql.opt_rule_blacklist values("join_reorder"),("topn_push_down");
+```
+
+> **注意：**
+> TiDB 不会实时地将这个表读入到内存中，所以即便是写入了禁用规则也不会立马生效。
+
+然后用户需要执行以下 SQL，让 TiDB 把禁用规则写到内存中，才能真正做到禁用的效果：
```
Comment (eurekaka) [inline reference/performance/opt-rule-blacklist.md:57]:
```suggestion
然后用户需要执行以下 SQL，让 TiDB 把禁用规则加载到缓存中，才能真正做到禁用的效果：
```

File: reference/performance/opt-rule-blacklist.md [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+当用户想禁用某些规则时，可以往 `mysql.opt_rule_blacklist` 的表中写入规则的名字，比如：
+
+{{< copyable "sql" >}}
+
+```sql
+insert into mysql.opt_rule_blacklist values("join_reorder"),("topn_push_down");
+```
+
+> **注意：**
+> TiDB 不会实时地将这个表读入到内存中，所以即便是写入了禁用规则也不会立马生效。
```
Comment (eurekaka) [inline reference/performance/opt-rule-blacklist.md:55]:
```suggestion
> TiDB 不会实时地将这个表读入到缓存中，所以即便是写入了禁用规则也不会立马生效。
```

File: reference/performance/opt-rule-blacklist.md [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+然后用户需要执行以下 SQL，让 TiDB 把禁用规则写到内存中，才能真正做到禁用的效果：
+
+{{< copyable "sql" >}}
+
+```sql
+admin reload opt_rule_blacklist
+```
+
+同理，需要解除某些规则的禁用时，需要执行删除表中数据并执行 `admin reload`:
```
Comment (eurekaka) [inline reference/performance/opt-rule-blacklist.md:65]:
```suggestion
同理，需要解除某些规则的禁用时，需要删除表中数据再执行 `admin reload`:
```

File: reference/performance/opt-rule-blacklist.md [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+
+{{< copyable "sql" >}}
+
+```sql
+delete from mysql.opt_rule_blacklist where name in ("join_reoder", "topn_push_down")
+admin reload opt_rule_blacklist
+```
+
+> **注意：**
+> 所以以上操作需要使用 root 权限的用户进行操作。
```
Comment (eurekaka) [inline reference/performance/opt-rule-blacklist.md:75]:
```suggestion
> 以上操作都需要数据库的 root 权限。
```

File: reference/performance/opt-rule-blacklist.md [hunk 14]
```diff
@@ -0,0 +1,75 @@
+---
+title: 优化规则黑名单
+category: reference
+---
+
+# 优化规则黑名单
+
+在[join reorder 算法简介](/reference/performance/join-reorder.md)中，介绍了一种较为复杂的优化规则，TiDB 对于某条查询，会智能地选择一些优化规则对查询进行智能优化。
+
+同时 TiDB 提供了一些可以自定义优化的功能，方便用户针对查询进行调优，比如[优化器 Hints](/reference/performance/optimizer-hints.md)。
```
Comment (eurekaka) [inline reference/performance/opt-rule-blacklist.md:10]: 上面 join reorder 和这一句都和当前文档没啥关系？要不删掉？

File: reference/performance/opt-rule-blacklist.md (hunk 6 above)
Comment (lzmhhh123) [reply reference/performance/opt-rule-blacklist.md:32]: 这里不止下推过 join，还有 union 等等。

File: reference/performance/opt-rule-blacklist.md [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+{{< copyable "sql" >}}
+
+```sql
+insert into mysql.opt_rule_blacklist values("join_reorder"),("topn_push_down");
+```
+
+> **注意：**
+> TiDB 不会实时地将这个表读入到缓存中，所以即便是写入了禁用规则也不会立马生效。
+
+然后用户需要执行以下 SQL，让 TiDB 把禁用规则加载到缓存中，才能真正做到禁用的效果：
```
Comment (Deardrops) [inline reference/performance/opt-rule-blacklist.md:57]:
```suggestion
执行以下 SQL 让禁用规则立即生效：
```

File: reference/performance/opt-rule-blacklist.md [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+
+{{< copyable "sql" >}}
+
+```sql
+delete from mysql.opt_rule_blacklist where name in ("join_reoder", "topn_push_down")
+admin reload opt_rule_blacklist
+```
+
+> **注意：**
+> 所以以上操作需要使用 root 权限的用户进行操作。
```
Comment (Deardrops) [inline reference/performance/opt-rule-blacklist.md:75]: 可以将这个提示移动到第一条的 SQL 的前面。

File: reference/performance/opt-rule-blacklist.md [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+
+- join 重排序（join_reorder）：对多表 join 确定连接顺序。
+
+## 如何禁用优化规则
+
+当某些优化规则对于一些比较特殊的查询，出现了不理想的优化结果时，用户可以使用**优化规则黑名单**对一些优化规则进行禁用。
+
+### 使用方法
+
+每个优化规则都有自己的名字，比如列裁剪的名字是 "column_prune"，所有优化规则的名字皆可以在[重要的优化规则](/reference/performance/opt-rule-blacklist.md#rules)中查到。
```
Comment (Deardrops) [inline reference/performance/opt-rule-blacklist.md:44]: 最好将这些规则做成表格形式，可读性好，方便复制。

Comment (zz-jason) [reply reference/performance/opt-rule-blacklist.md:44]: +1

File: reference/performance/opt-rule-blacklist.md [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+
+# 优化规则黑名单
+
+**优化规则黑名单**是针对优化规则的调优手段之一，主要用于手动禁用一些优化规则。
+
+## 重要的优化规则<div id="rules"></div>
+
+|**优化规则**|**规则名称**|**简介**|
+| --- | --- | --- |
+| 列裁剪 | column_prune | 对于上层算子不需要的列，在下层算子不输出该列，减少计算 |
```
Comment (TomShawn) [inline reference/performance/opt-rule-blacklist.md:14]:
```suggestion
| 列裁剪 | column_prune | 对于上层算子不需要的列，不在下层算子输出该列，减少计算 |
```

File: reference/performance/opt-rule-blacklist.md [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+| 子查询去关联 | decorrelate | 尝试对相关子查询进行改写，将其转换为普通 join 或 aggression 计算 |
+| 聚合消除 | aggregation_eliminate | 尝试消除执行计划中的某些不必要的聚合算子 |
+| 投影消除 | projection_eliminate | 对于执行计划中不必要的投影算子，对其消除 |
```
Comment (TomShawn) [inline reference/performance/opt-rule-blacklist.md:17]:
```suggestion
| 投影消除 | projection_eliminate |  消除执行计划中不必要的投影算子 |
```

File: reference/performance/opt-rule-blacklist.md [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+| 最大最小消除 | max_min_eliminate | 改写聚合中的 max, min 计算，转化为 order by + limit 1 |
```
Comment (TomShawn) [inline reference/performance/opt-rule-blacklist.md:18]:
```suggestion
| 最大最小消除 | max_min_eliminate | 改写聚合中的 max/min 计算，转化为 `order by` + `limit 1` |
```

File: reference/performance/opt-rule-blacklist.md [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+| 谓词下推 | predicate_push_down | 尝试将执行计划中过滤条件下推到离数据源更近的算子上 |
+| 外连接消除 | outer_join_eliminate | 尝试消除执行计划中不必要的 left join 或者 right join |
+| 分区裁剪 | partition_processor | 将分区表查询改成为用 union all，并裁剪掉不满足过滤条件的分区 |
+| 聚合下推 | aggregation_push_down | 尝试将执行计划中的聚合算子下推到更底层的计算节点 |
+| TopN 下推 | topn_push_down | 尝试将执行计划中的 TopN 算子下推到离数据源更近的算子上 |
+| Join 重排序 | join_reorder | 对多表 join 确定连接顺序 |
+
+## 如何禁用优化规则
+
+当某些优化规则对于一些比较特殊的查询，出现了不理想的优化结果时，用户可以使用**优化规则黑名单**对一些优化规则进行禁用。
```
Comment (TomShawn) [inline reference/performance/opt-rule-blacklist.md:28]: `比较特殊的查询`可否举个例子？
```suggestion
当某些优化规则在一些特殊查询中的优化结果不理想时，用户可以使用**优化规则黑名单**禁用一些优化规则。
```

File: reference/performance/opt-rule-blacklist.md [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+
+### 使用方法
+
+> **注意：**
```
Comment (TomShawn) [inline reference/performance/opt-rule-blacklist.md:32]:
```suggestion
> **注意：**
>
```

File: reference/performance/opt-rule-blacklist.md [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+> 以下操作都需要数据库的 root 权限。
+
+每个优化规则都有自己的名字，比如列裁剪的名字是 "column_prune"，所有优化规则的名字皆可以在[重要的优化规则](/reference/performance/opt-rule-blacklist.md#rules)表格中第二列查到。
```
Comment (TomShawn) [inline reference/performance/opt-rule-blacklist.md:35]:
```suggestion
每个优化规则都有各自的名字，比如列裁剪的名字是 "column_prune"。所有优化规则的名字都可以在[重要的优化规则](#重要的优化规则)表格中第二列查到。
```

File: reference/performance/opt-rule-blacklist.md [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+
+当用户想禁用某些规则时，可以往 `mysql.opt_rule_blacklist` 的表中写入规则的名字，比如：
```
Comment (TomShawn) [inline reference/performance/opt-rule-blacklist.md:37]:
```suggestion
当用户想禁用某些规则时，可以在 `mysql.opt_rule_blacklist` 表中写入规则的名字，比如：
```

File: reference/performance/opt-rule-blacklist.md [hunk 25]
```diff
@@ -0,0 +1,60 @@
...
+
+{{< copyable "sql" >}}
+
+```sql
+insert into mysql.opt_rule_blacklist values("join_reorder"),("topn_push_down");
+```
+
+执行以下 SQL 让禁用规则立即生效：
```
Comment (TomShawn) [inline reference/performance/opt-rule-blacklist.md:45]:
```suggestion
执行以下 SQL 语句让禁用规则立即生效：
```

File: reference/performance/opt-rule-blacklist.md [hunk 26]
```diff
@@ -0,0 +1,60 @@
...
+
+{{< copyable "sql" >}}
+
+```sql
+admin reload opt_rule_blacklist
+```
+
+同理，需要解除某些规则的禁用时，需要删除表中数据再执行 `admin reload`:
```
Comment (TomShawn) [inline reference/performance/opt-rule-blacklist.md:53]:
```suggestion
需要解除一条规则的禁用时，需要删除表中禁用该条规则的相应数据，再执行 `admin reload`：
```

File: reference/performance/opt-rule-blacklist.md [hunk 27]
```diff
@@ -0,0 +1,60 @@
+---
+title: 优化规则黑名单
+category: reference
+---
+
+# 优化规则黑名单
+
+**优化规则黑名单**是针对优化规则的调优手段之一，主要用于手动禁用一些优化规则。
+
+## 重要的优化规则<div id="rules"></div>
```
[the remark this answers is not in the document]
Comment (lzmhhh123) [reply reference/performance/opt-rule-blacklist.md:10]: This <div> label helps the following link to redirect.

[the remark this answers is not in the document]
Comment (zz-jason) [reply reference/performance/opt-rule-blacklist.md:10]: can it be rendered correctly in the web page?

[the remark this answers is not in the document]
Comment (lzmhhh123) [reply reference/performance/opt-rule-blacklist.md:10]: As I know, it works in gitbook. Even if it doesn't work in our document, there are none other effects.

File: reference/performance/opt-rule-blacklist.md [hunk 28]
```diff
@@ -0,0 +1,61 @@
...
+
+## 如何禁用优化规则
+
+当某些优化规则在一些特殊查询中的优化结果不理想时，用户可以使用**优化规则黑名单**禁用一些优化规则。
+
+### 使用方法
+
+> **注意：**
+>
+> 以下操作都需要数据库的 root 权限。
```
Comment (zz-jason) [inline reference/performance/opt-rule-blacklist.md:34]: should it be `super privilege`? please use the word in the privilege system.

File: reference/performance/opt-rule-blacklist.md [hunk 29]
```diff
@@ -0,0 +1,61 @@
...
+category: reference
+---
+
+# 优化规则黑名单
+
+**优化规则黑名单**是针对优化规则的调优手段之一，主要用于手动禁用一些优化规则。
+
+## 重要的优化规则
+
+|**优化规则**|**规则名称**|**简介**|
```
Comment (winoros) [inline reference/performance/opt-rule-blacklist.md:12]: 找个地方指回 `SQL优化流程` 吧。

File: optimization-rule-blacklist.md [hunk 30]
```diff
@@ -0,0 +1,61 @@
...
+
+## 如何禁用优化规则
+
+当某些优化规则在一些特殊查询中的优化结果不理想时，用户可以使用**优化规则黑名单**禁用一些优化规则。
+
+### 使用方法
+
+> **注意：**
+>
+> 以下操作都需要数据库的 root 权限。
```
Comment (zz-jason) [inline optimization-rule-blacklist.md:34]: 确认下是 super privilege 还是 root 权限呢，MySQL 没有 root 权限：https://dev.mysql.com/doc/refman/8.0/en/privileges-provided.html

File: optimization-rule-blacklist.md [hunk 31]
```diff
@@ -0,0 +1,61 @@
...
+
+每个优化规则都有各自的名字，比如列裁剪的名字是 "column_prune"。所有优化规则的名字都可以在[重要的优化规则](#重要的优化规则)表格中第二列查到。
+
+当用户想禁用某些规则时，可以在 `mysql.opt_rule_blacklist` 表中写入规则的名字，比如：
```
Comment (zz-jason) [inline optimization-rule-blacklist.md:38]: 举一个 SQL 的执行计划变化的例子向用户多解释下如何确认优化规则生效了吧

**Pull request closed without merging**